### PR TITLE
Assign requisition_number and invoice_number to transfers

### LINKED ISSUE #1239: Stock transfer invoice
## What went wrong? 😲

Stock transfer either by outbound invoice or via internal order request > response, Viewing inbound shipment in remote client, complaints says `Int parsing error`.

invoice -1

## Expected behaviour 🤔

Not sure what is the idea behind, probably central server logic need changing or the remote client should be handling it.

## How to Reproduce 🔨

Steps to reproduce the behaviour:

1. Do either, normal stock transfer or request, response order work flow
2. Once the response is sent from responder (Internal order work flow), or outbound invoice is finalised > Creates its associated Supplier invoice
3. Sync remote client
4. See inbound shipment
5. Click to view details
6. Int parse error

## Your environment 🌱
 - Version [1.2.3 , current develop]
 - Platform [browser, windows desktop, server]

### PR DESCRIPTION
**Pull request by mark-prins**
Fixes #1239

# 👩🏻‍💻 What does this PR do? 
 
When a response requisition or an inbound shipment is created automatically by the central server, if the server is unable to determine the next number in the sequence for the destination store, then it simply assigns `-1` on the expectation that the store will allocate the number when processing the transfer.

omSupply wasn't allocating - resulting in `-1` for the `invoice_number` and `requisition_number` - this is a problem as we are using the number to lookup the shipment / requisition.

This change is to watch out for `-1` and allocate a number if required.

In addition - the response requisition was not mapping status from `sg` which resulted in an error when processing. This then causes problems - because the linked requisition doesn't exist, the requisition processing is blocked.

# 🧪 How has/should this change been tested? 

Create an outbound shipment in an mSupply store, sending stock to an omSupply store. Add a line and finalise.
This will create an inbound shipment for the destination store. Check that this has a valid invoice_number.

Similarly - create an internal order in an mSupply store, requesting stock from an omSupply store.

## 💌 Any notes for the reviewer?

Have added a log entry when the number is allocated. For example, with a requisition:

Screenshot 2023-04-05 at 4 02 13 PM

## 📃 Documentation

_No user facing changes_

### CODE REVIEW

File: server/service/src/processors/transfer/shipment/create_inbound_shipment.rs [hunk 1]
```diff
@@ -63,7 +63,47 @@ impl ShipmentTransferProcessor for CreateInboundShipmentProcessor {
...
+                InvoiceRowRepository::new(connection).upsert_one(&updated_invoice_row)?;
+                system_activity_log_entry(
+                    connection,
+                    ActivityLogType::InvoiceNumberAllocated,
+                    &inbound_shipment.store_row.id,
+                    &inbound_shipment.invoice_row.id,
+                )?;
+            }
+
+            // generate lines for the inbound shipment
```
Comment (mark-prins) [inline server/service/src/processors/transfer/shipment/create_inbound_shipment.rs:87]: the inbound shipment had no rows, so I had to add them 🤷

File: server/service/src/sync/translations/requisition.rs [hunk 2]
```diff
@@ -382,7 +382,7 @@ fn from_legacy_status(
             _ => return None,
         },
         LegacyRequisitionType::Response => match status {
-            LegacyRequisitionStatus::Sg => return None,
+            LegacyRequisitionStatus::Sg => RequisitionRowStatus::New,
```
Comment (mark-prins) [inline server/service/src/sync/translations/requisition.rs:385]: returning `None` meant that the fn which is mapping the status raises an error. This prevents processing of requisitions.

**Review by andreievg — CHANGES_REQUESTED** [review]
Thanks @mark-prins, at first I was going to suggest that we don't do this and let mSupply fix this by not creating transfer for omSupply site, but then realised this code would be needed in the future when we migrate for mSupply sites.

I still think we need to make an issue for:
* mSupply to not generate transfers for omSupply site when requesting site is active on central, i think it's this one: https://github.com/openmsupply/open-msupply/issues/1396
* When mSupply does generate supplier invoice (for mSupply site) to not duplicate item_line_id of customer invoice line in supplier invoice line (otherwise these will fail to integrate in supplier invoice site), btw this would be part of Migration work

The requested changes are (as per comment)
* Separate processors
* New creation of invoice lines would create duplicates

**Comment by mark-prins** [comment]
thanks @andreievg changes made in 6f1581415
Have removed the creation of lines - so that the lines are not duplicated in mSupply and do not exist in omSupply.
Creating an issue on mSupply repo to fix that now.

**Comment by mark-prins** [comment]
Issue [#1512](https://github.com/openmsupply/open-msupply/issues/1512) created

File: server/service/src/sync/translations/invoice_line.rs [hunk 3]
```diff
@@ -153,14 +153,25 @@ impl SyncTranslation for InvoiceLineTranslation {
...
+
+        let is_record_active_on_site = is_active_record_on_site(
             &connection,
             ActiveRecordCheck::InvoiceLine {
                 invoice_id: invoice_id.clone(),
             },
-        )? {
+        )?;
+
+        let is_stock_line_valid = match stock_line_id {
```
Comment (andreievg) [inline server/service/src/sync/translations/invoice_line.rs:164]: Hmm, i don't actually agree with this change, I'll point it out in the issue you've created for mSupply (to get this PR merged), I think not having stock line referenced by active record on site should be logged and not silently ignored (previously row would have been missing but logged against sync_buffer)

**Review by andreievg — APPROVED** [review]
Thanks @mark-prins, please see comment I've added: https://github.com/sussol/msupply/issues/12729#issuecomment-1505306848

**Comment by mark-prins** [comment]
🙏  thanks @andreievg 
have added a warning to the log, and a TODO to take out the stock_line_id change.

**Pull request merged**